File: LeadToOffer/24_PrintPathSumInTree.py
## recursively find down.
class Solution:
    def FindPath(self, root, expectNumber):
        paths = []
        if root is None:
            return []

        if root.left is None and root.right is None and root.val==expectNumber:
            paths.append([expectNumber])

        if root.left is not None:
            temp = self.FindPath(root.left, expectNumber - root.val)
            if len(temp)!=0 and len(temp[0])!=0:
                for v in temp:
                    v = [root.val] + v
                    paths.append(v)
        if root.right is not None:
            temp = self.FindPath(root.right, expectNumber - root.val)
            if len(temp)!=0 and len(temp[0])!=0:
                for v in temp:
                    v = [root.val] + v
                    paths.append(v)
        return paths

File: LeadToOffer/24_PrintPathSumInTree.cpp
class Solution {
public:
    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {
        vector<vector<int> > paths;
        if (!root)
            return paths;

        if (!root->left && !root->right && root->val == expectNumber) {
            vector<int> single;
            single.push_back(expectNumber);
            paths.push_back(single);
        }

        vector<vector<int> > temp;
        vector<int> vec;
        if (root->left) {
            temp = FindPath(root->left, expectNumber - root->val);
            if (temp.size()!=0 && temp[0].size()!=0) {
                for(int i=0; i<temp.size(); i++){
                    vec = temp[i];
                    vec.insert(vec.begin(), root->val);
                    paths.push_back(vec);
                    vec.clear();
                }
            }
        }
        if (root->right) {
            temp = FindPath(root->right, expectNumber - root->val);
            if (temp.size()!=0 && temp[0].size()!=0) {
               for(int i=0; i<temp.size(); i++){
                    vec = temp[i];
                    vec.insert(vec.begin(), root->val);
                    paths.push_back(vec);
                    vec.clear();
                }
            }
        }

        return paths;
    }
};
